### ISSUE TITLE
System.ArgumentException when using CreateProgram with string[] as source parameter

### ISSUE THREAD

**Issue by CamelCaseName** [dotnet/Silk.NET#1264]
# Summary
calling 
```cs
cl.CreateProgramWithSource(nint context, uint count, string[] stringsSa, nuint* lengths, out int errcode_ret);
``` 
with a valid context and valid kernel as a string in Kernels.LayoutKernel throws a "System.ArgumentException" from within the string array to pointer conversion code. The Kernel is about 4500 characters long, can be found [here](https://github.com/CamelCaseName/HousePartyTranslator/blob/2.0_web_rework/TranslatorAdmin/Explorer/OpenCL/Layout.cl). providing a valid length didn't help, and is not needed per se if the string is null terminated according to the OpenCL docs.

# Stacktrace
```
System.ArgumentException
  HResult=0x80070057
  Message=The output byte buffer is too small to contain the encoded data, encoding codepage '65001' and fallback 'System.Text.EncoderReplacementFallback'. Arg_ParamName_Name
  Source=System.Private.CoreLib
  StackTrace:
   at System.Text.Encoding.ThrowBytesOverflow() in /_/src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs:line 1090
   at System.Text.Encoding.ThrowBytesOverflow(EncoderNLS encoder, Boolean nothingEncoded) in /_/src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs:line 1101
   at System.Text.Encoding.GetBytesWithFallback(ReadOnlySpan`1 chars, Int32 originalCharsLength, Span`1 bytes, Int32 originalBytesLength, EncoderNLS encoder) in /_/src/libraries/System.Private.CoreLib/src/System/Text/Encoding.Internal.cs:line 687
   at System.Text.Encoding.GetBytesWithFallback(Char* pOriginalChars, Int32 originalCharCount, Byte* pOriginalBytes, Int32 originalByteCount, Int32 charsConsumedSoFar, Int32 bytesWrittenSoFar) in /_/src/libraries/System.Private.CoreLib/src/System/Text/Encoding.Internal.cs:line 497
   at System.Text.UTF8Encoding.GetBytes(Char* chars, Int32 charCount, Byte* bytes, Int32 byteCount) in /_/src/libraries/System.Private.CoreLib/src/System/Text/UTF8Encoding.cs:line 358
   at Silk.NET.Core.Native.SilkMarshal.StringIntoSpan(String input, Span`1 span, NativeStringEncoding encoding)
   at Silk.NET.Core.Native.SilkMarshal.StringToMemory(String input, NativeStringEncoding encoding)
   at Silk.NET.Core.Native.SilkMarshal.StringToPtr(String input, NativeStringEncoding encoding)
   at Silk.NET.Core.Native.SilkMarshal.StringArrayToMemory(IReadOnlyList`1 input, NativeStringEncoding e)
   at Silk.NET.Core.Native.SilkMarshal.StringArrayToPtr(IReadOnlyList`1 input, NativeStringEncoding encoding)
   at Silk.NET.OpenCL.CL.CreateProgramWithSource(IntPtr context, UInt32 count, String[] stringsSa, UIntPtr* lengths, Int32& errcode_ret)
   at Translator.Explorer.OpenCL.OpenCLManager.CreateProgram() in F:\Lenny\source\repos\HousePartyTranslator\TranslatorAdmin\Explorer\OpenCL\OpenCLManager.cs:line 74
   at Translator.Explorer.OpenCL.OpenCLManager.SetUpOpenCLImpl() in F:\Lenny\source\repos\HousePartyTranslator\TranslatorAdmin\Explorer\OpenCL\OpenCLManager.cs:line 296
   at Translator.Explorer.OpenCL.OpenCLManager.SetUpOpenCL() in F:\Lenny\source\repos\HousePartyTranslator\TranslatorAdmin\Explorer\OpenCL\OpenCLManager.cs:line 46
   at Translator.Explorer.NodeLayout..ctor(NodeProvider provider, Form parent, CancellationToken cancellation) in F:\Lenny\source\repos\HousePartyTranslator\TranslatorAdmin\Explorer\NodeLayout.cs:line 51
   at Translator.Explorer.Window.StoryExplorer.Initialize(Boolean singleFile) in F:\Lenny\source\repos\HousePartyTranslator\TranslatorAdmin\Explorer\StoryExplorer.cs:line 77
   at Translator.Fenster.<>c__DisplayClass80_1.<CreateStoryExplorer>b__0() in F:\Lenny\source\repos\HousePartyTranslator\TranslatorAdmin\MainForm\House Party Translator.cs:line 362
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ExecutionContext.cs:line 268

  This exception was originally thrown at this call stack:
    System.Text.Encoding.ThrowBytesOverflow() in Encoding.cs
    System.Text.Encoding.ThrowBytesOverflow(System.Text.EncoderNLS, bool) in Encoding.cs
    System.Text.Encoding.GetBytesWithFallback(System.ReadOnlySpan<char>, int, System.Span<byte>, int, System.Text.EncoderNLS) in Encoding.Internal.cs
    System.Text.Encoding.GetBytesWithFallback(char*, int, byte*, int, int, int) in Encoding.Internal.cs
    System.Text.UTF8Encoding.GetBytes(char*, int, byte*, int) in UTF8Encoding.cs
    Silk.NET.Core.Native.SilkMarshal.StringIntoSpan(string, System.Span<byte>, Silk.NET.Core.Native.NativeStringEncoding)
    Silk.NET.Core.Native.SilkMarshal.StringToMemory(string, Silk.NET.Core.Native.NativeStringEncoding)
    Silk.NET.Core.Native.SilkMarshal.StringToPtr(string, Silk.NET.Core.Native.NativeStringEncoding)
    Silk.NET.Core.Native.SilkMarshal.StringArrayToMemory(System.Collections.Generic.IReadOnlyList<string>, Silk.NET.Core.Native.NativeStringEncoding)
    Silk.NET.Core.Native.SilkMarshal.StringArrayToPtr(System.Collections.Generic.IReadOnlyList<string>, Silk.NET.Core.Native.NativeStringEncoding)
    ...
    [Call Stack Truncated]
```
# Workaround
using a `byte[]` works, which can be easily achieved like so
```cs
byte[] codeBytes = new byte[Encoding.Default.GetByteCount(Kernels.LayoutKernel) + 1];
Encoding.Default.GetBytes(Kernels.LayoutKernel, 0, Kernels.LayoutKernel.Length, codeBytes, 0);
fixed (byte* code = codeBytes)
{
	_program = _cl.CreateProgramWithSource(_context, 1, in code, null, out err);
	if (err != 0) return err;
}
```

# Steps to reproduce
- Platform: Desktop
- Framework Version: .NET 7.0
- API: OpenCL (Silk.NET.OpenCL v 2.16.0)
- API Version: OpenCL 2.2? 3.0? no idea how to find out
1. create a valid context
2. get a device pointer
3. call the function above with valid parameters

# Comments
if you need the rest of the setup code, its in the repo linked with the kernel

**Comment by Perksey** [comment]
I've fixed this issue in an upcoming PR.

I unfortunately couldn't reproduce. I added [this file](https://github.com/CamelCaseName/HousePartyTranslator/blob/07117f1/TranslatorAdmin/Explorer/OpenCL/NBody.cl)'s contents (perma link because original link was dead) as a single string constant, and passed a string array of length 1 containing that constant to SilkMarshal.StringArrayToPtr.

However, I did manage to find what I think was the bug. For a while now we always encode LPStrs as if they were LPUTF8Strs, but didn't account for this in GetMaxSizeOf. The calculation for LPStr can vary from system to system. As such, I've just changed it to use the proper calculation which will fix your issue.

**Comment by CamelCaseName** [comment]
Oh very nice, your plausible cause of the bug is also what I found when I stepped through the conversion, the lengths were off by a few bytes iirc. "Fixed" it on my end with the proposed workaround, but thanks anyways for maintaining!

**Comment by Beyley** [comment]
Closed by #1372
